include/alpaka/atomic/AtomicUniformCudaHipBuiltIn.hpp

```diff
@@ -185,14 +185,70 @@ namespace alpaka
-                    static_assert(
-                        !sizeof(THierarchy),
-                        "atomicOp<AtomicSub, AtomicUniformCudaHipBuiltIn, unsigned long int> is only supported when "
-                        "sizeof(unsigned long int) == 4");
+                {
+                    return ::atomicAdd(
+                        reinterpret_cast<unsigned long long int*>(addr),
+                        static_cast<unsigned long long int>(-value));
+                }
 
                 ALPAKA_UNREACHABLE(0ul);
             }
         };
+        //! The GPU CUDA/HIP accelerator atomic operation.
+        template<typename THierarchy>
+        struct AtomicOp<AtomicSub, AtomicUniformCudaHipBuiltIn, unsigned long long int, THierarchy>
+        {
+            //
+            __device__ static auto atomicOp(
+                AtomicUniformCudaHipBuiltIn const&,
+                unsigned long long int* const addr,
+                unsigned long long int const& value) -> unsigned long long int
+            {
+                return ::atomicAdd(addr, -value);
+            }
+        };
+        //! The GPU CUDA/HIP accelerator atomic operation.
+        template<typename THierarchy>
+        struct AtomicOp<AtomicSub, AtomicUniformCudaHipBuiltIn, float, THierarchy>
+        {
+            //
+            __device__ static auto atomicOp(AtomicUniformCudaHipBuiltIn const&, float* const addr, float const& value)
+                -> float
+            {
+                return ::atomicAdd(addr, -value);
+            }
+        };
+        //! The GPU CUDA/HIP accelerator atomic operation.
+        template<typename THierarchy>
+        struct AtomicOp<AtomicSub, AtomicUniformCudaHipBuiltIn, double, THierarchy>
+        {
+            __device__ static auto atomicOp(
+                AtomicUniformCudaHipBuiltIn const&,
+                double* const addr,
+                double const& value) -> double
+            {
+#        if BOOST_ARCH_PTX >= BOOST_VERSION_NUMBER(6, 0, 0)
+                return ::atomicAdd(addr, -value);
+#        else
+                // Code from: http://docs.nvidia.com/cuda/cuda-c-programming-guide/#atomic-functions
+
+                unsigned long long int* address_as_ull(reinterpret_cast<unsigned long long int*>(addr));
+                unsigned long long int old(*address_as_ull);
+                unsigned long long int assumed;
+                do
+                {
+                    assumed = old;
+                    old = ::atomicCAS(
+                        address_as_ull,
+                        assumed,
+                        static_cast<unsigned long long>(
+                            __double_as_longlong(-value + __longlong_as_double(static_cast<long long>(assumed)))));
+                    // Note: uses integer comparison to avoid hang in case of NaN (since NaN != NaN)
+                } while(assumed != old);
+                return __longlong_as_double(static_cast<long long>(old));
+#        endif
```

Comment:
```suggestion
            return ::atomicAdd(addr, -value);
```
I don't think we need to duplicate the CAS workaround here. `atomicAdd` can do that.

Reply:
I tried but it gives me an error:
``` 
/data/user/adipilat/alpaka/include/alpaka/atomic/AtomicUniformCudaHipBuiltIn.hpp(230): error: no instance of overloaded function "atomicAdd" matches the argument list
            argument types are: (double *const, double)
```

I also tried to to that for atomicCAS when implementing support for `unsigned long int` (the native atomicCAS does not support the type, but there's an alpaka implementation that does it) instead of recasting types.

Reply:
Oh wait, it needs to be `atomicAdd`, not `::atomicAdd`. Or alternatively:
`AtomicOp<AtomicAdd, AtomicUniformCudaHipBuiltIn, double, THierarchy>::atomicOp(..., addr, -value);`

You would call the alpaka one, not the CUDA one directly.

Reply:
`atomicAdd` gives the same error, although it's clearly the alpaka one being called (```/data/user/adipilat/alpaka/include/alpaka/atomic/AtomicUniformCudaHipBuiltIn.hpp(233): error: no instance of function template "alpaka::atomicAdd" matches the argument list
            argument types are: (double *const, double)
```

Reply:
Because you are missing the first argument that alpaka's version needs. Here is the complete version:
```c++
        template<typename THierarchy>
        struct AtomicOp<AtomicSub, AtomicUniformCudaHipBuiltIn, double, THierarchy>
        {
            __device__ static auto atomicOp(
                AtomicUniformCudaHipBuiltIn const& ctx,
                double* const addr,
                double const& value) -> double
            {
                atomicAdd(ctx, addr, -value);
            }
        };
```